Add support for persistent SCC on z/OS

hangshao0 opened the pull request:
1. Use mmap implementation of SCC on z/OS which can be enabled by
"persistent" sub-option.
2. Flush the change of the mapped region to the SCC file on VM exit.
3. Use map64 by default on Java 11 and up. Add new sub-option "map31"
which maps SCC below the 2G addressing range.
4. Add test cases
5. Remove the temporary define of OMRPORT_MMAP_FLAG_ZOS_READ_MAPFILE 

Depends on https://github.com/eclipse/omr/pull/6940
Depends on https://github.com/eclipse/omr/pull/6989

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

hangshao0 commented:
@pshipton

keithc-ca commented on runtime/nls/shrc/j9shr.nls:
```diff
@@ -6896,3 +6896,24 @@ J9NLS_SHRC_CM_PRINTSTATS_PROCESSOR_FEATURES.explanation=NOTAG
+J9NLS_SHRC_OSCACHE_MMAP_MSYNC_ERROR=Error Synchronizing memory
+# START NON-TRANSLATABLE
+J9NLS_SHRC_OSCACHE_MMAP_MSYNC_ERROR.explanation=An error has occurred in shared class processing.
+J9NLS_SHRC_OSCACHE_MMAP_MSYNC_ERROR.system_action=The JVM terminates, unless you have specified the nonfatal option with "-Xshareclasses:nonfatal", in which case the JVM continues without using Shared Classes.
```
I think the first occurrence of "nonfatal" should be hyphenated, i.e. "non-fatal" and "Shared Classes" need not be capitalized.

hangshao0 replied:
Fixed.

keithc-ca commented on runtime/oti/j9port.h:
```diff
@@ -965,6 +965,9 @@ typedef struct J9CacheInfoQuery {
 #define J9PORT_MMAP_FLAG_SHARED OMRPORT_MMAP_FLAG_SHARED
 #define J9PORT_MMAP_FLAG_COPYONWRITE OMRPORT_MMAP_FLAG_COPYONWRITE
 #define J9PORT_MMAP_SYNC_WAIT OMRPORT_MMAP_SYNC_WAIT
+#if defined(J9ZOS39064)
+#define J9PORT_MMAP_FLAG_ZOS_64BIT OMRPORT_MMAP_FLAG_ZOS_64BIT
+#endif /* defined(J9ZOS39064) */
```
Why shouldn't we just use `OMRPORT_MMAP_FLAG_ZOS_64BIT` instead of creating an alias?

hangshao0 replied:
We can use `OMRPORT_MMAP_FLAG_ZOS_64BIT`. But I see we are using other J9PORT_MMAP_FLAG_XXX in existing OpenJ9 code, so I add `J9PORT_MMAP_FLAG_ZOS_64BIT`. This way we don't need to check and mix use `J9PORT_` version for some flags and `OMRPORT_` versions for others.

pshipton commented on runtime/shared_common/CompositeCache.cpp:
```diff
@@ -1396,32 +1381,61 @@ SH_CompositeCacheImpl::startup(J9VMThread* currentThread, J9SharedClassPreinitCo
 
-#if defined (J9SHR_MSYNC_SUPPORT)
+#if defined(J9SHR_MSYNC_SUPPORT)
 				_doHeaderSync = _doReadWriteSync = _doSegmentSync = _doMetaSync = (*_runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_MSYNC);
```
This can move into the previous #if defined(J9SHR_MSYNC_SUPPORT)

hangshao0 replied:
The previous `#if defined(J9SHR_MSYNC_SUPPORT)` is inside `if` condition.

pshipton commented on test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml:
```diff
@@ -1426,6 +1426,45 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 	</test>
 
+
+	<test id="Test 67-a: Test classes are being stored to the persisnet shared cache" timeout="600" runPath=".">
```
typo persisnet, here and below

hangshao0 replied:
Fixed.

hangshao0 commented:
> What are the concerns with making map64 the default?

`__MAP_64` is only supported on z/OS version 2.5 and some z/OS version 2.4. The 64 bit mapping is not on by default on z/OS `mmap()` where it is supported. Both @r30shah and I have seen  `__MAP_64` is easy to cause out of memory error (`ENOMEM`) on `mmap()`. The `ENOMEM` goes away if a smaller cache size is used. There might be some machine configurations that could control this. Not sure if @r30shah has found out something on this.

> The testing is light. Have you tried making a JVM with persistent as the default and running all the functional tests?

Not yet.

hangshao0 commented on runtime/bcutil/jimagereader.c:
```diff
@@ -179,8 +180,13 @@ j9bcutil_loadJImage(J9PortLibrary *portlib, const char *fileName, J9JImage **pji
-	jimage->jimageMmap = j9mmap_map_file(jimagefd, 0, mapSize, fileName, J9PORT_MMAP_FLAG_READ, J9MEM_CATEGORY_CLASSES);
+#if defined(J9ZOS390)
+	/*
+	 * With OMRPORT_MMAP_FLAG_ZOS_READ_MAPFILE, j9mmap_map_file() has the old bahaviour that reads the file content into allocate private memory.
```
Fixed.

hangshao0 commented:
I've also added change regarding:
https://github.com/eclipse/omr/pull/6940#issuecomment-1490862402.

hangshao0 commented:
PR for the OMR flag https://github.com/eclipse-openj9/openj9/pull/17083

hangshao0 commented:
As https://github.com/eclipse-openj9/openj9/pull/17083 is merged, I pushed to resolve the merge conflicts.

pshipton commented:
Now this PR can be updated to remove `#define OMRPORT_MMAP_FLAG_ZOS_READ_MAPFILE 0` because this won't be merged until after the OMR change is merged.

hangshao0 commented:
> Now this PR can be updated to remove #define OMRPORT_MMAP_FLAG_ZOS_READ_MAPFILE 0

Yes, it can be done.

keithc-ca commented on runtime/shared_common/OSCachemmap.cpp:
```diff
@@ -183,7 +183,9 @@ SH_OSCachemmap::startup(J9JavaVM* vm, const char* ctrlDirName, UDATA cacheDirPer
 	versionData->cacheType = J9PORT_SHR_CACHE_TYPE_PERSISTENT;
 	mmapCapabilities = j9mmap_capabilities();
-	if (!(mmapCapabilities & J9PORT_MMAP_CAPABILITY_WRITE) || !(mmapCapabilities & J9PORT_MMAP_CAPABILITY_MSYNC)) {
+	if (!(mmapCapabilities & J9PORT_MMAP_CAPABILITY_WRITE)
+		|| !(mmapCapabilities & J9PORT_MMAP_CAPABILITY_MSYNC)
```
This should use `J9_ARE_NO_BITS_SET()`:
```c
	if (J9_ARE_NO_BITS_SET(mmapCapabilities, J9PORT_MMAP_CAPABILITY_MSYNC | J9PORT_MMAP_CAPABILITY_WRITE)) {
```

hangshao0 replied:
Fixed.

keithc-ca commented on runtime/shared_common/shrinit.cpp:
```diff
@@ -277,6 +277,10 @@ J9SharedClassesHelpText J9SHAREDCLASSESHELPTEXT[] = {
 	{OPTION_DISABLE_CORRUPT_CACHE_DUMPS, 0, 0, J9NLS_SHRC_SHRINIT_HELPTEXT_DISABLE_CORRUPT_CACHE_DUMPS},
 	{OPTION_CHECK_STRINGTABLE_RESET, 0, 0, J9NLS_SHRC_SHRINIT_HELPTEXT_CHECK_STRINGTABLE_RESET},
 	{OPTION_ADDTESTJITHINT, 0, 0, J9NLS_SHRC_SHRINIT_HELPTEXT_ADD_JIT_HINTS},
+#if (defined(J9ZOS39064) && (JAVA_SPEC_VERSION >= 11))
```
Please remove redundant parentheses (should be like `#endif` comment). Same for line 383.

keithc-ca commented on runtime/tests/shared/AttachedDataTest.cpp:
```diff
@@ -201,8 +201,7 @@ AttachedDataTest::openTestCache(J9JavaVM* vm, BlockPtr preallocatedCache, UDATA
 	memset(piConfig, 0, sizeof(J9SharedClassPreinitConfig));
-	/* j9mmap_get_region_granularity returns 0 on zOS */
-	osPageSize = j9mmap_get_region_granularity(NULL);
+	osPageSize = (j9mmap_capabilities() & J9PORT_MMAP_CAPABILITY_PROTECT) ? j9mmap_get_region_granularity(NULL) : 0;
```
This should use `J9_ARE_ANY_BITS_SET()`.

keithc-ca commented on test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml:
```diff
@@ -1426,6 +1426,45 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<output type="success" caseSensitive="yes" regex="no" platforms=".*aarch64.*">Current statistics for cache</output>
 	</test>
 
+
```
Please remove this duplicate blank line.

hangshao0 replied:
Fixed.

hangshao0 commented on buildspecs/j9.flags:
```diff
@@ -1858,6 +1858,13 @@ Only available on zOS</description>
 			<require flag="interp_jniSupport"/>
 		</requires>
 	</flag>
+	<flag id="opt_shrMsyncSupport">
```
Currently this is enabled on z/OS only. Not sure about the future. I prefer to keep this flag for now just in case a second platform need to enable this in the future.

hangshao0 commented on runtime/shared_common/CompositeCache.cpp:
```diff
@@ -1044,7 +1021,15 @@ SH_CompositeCacheImpl::startup(J9VMThread* currentThread, J9SharedClassPreinitCo
 		setCurrentCacheVersion(vm, J2SE_VERSION(currentThread->javaVM), &versionData);
-
+#if defined(J9SHR_MSYNC_SUPPORT)
+		if (J9_ARE_ALL_BITS_SET(*_runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_PERSISTENT_CACHE)) {
```
Fixed.

hangshao0 commented on runtime/shared_common/CompositeCache.cpp:
```diff
@@ -3737,6 +3753,11 @@ SH_CompositeCacheImpl::runExitCode(J9VMThread *currentThread)
 		return;
 	}
+	/* If J9SHR_MSYNC_SUPPORT is enabled, syncUpdates() is needed to write the header page to the cache file.
+	 * call runExitCode() first so that the last detach time can be saved to the file.
```
Fixed.

hangshao0 commented:
> Have you tried making a JVM with persistent as the default and running all the functional tests?
> Actually the system tests as well, there are some class loading and shared cache related system tests.

The heavy class loading tests like `SharedClasses.SCM01.SingleCL`, `SharedClasses.SCM01.MultiThreadMultiCL` are in extended.system. 
But there are issues with Jenkins machines, no persistent SCC > 10m are allowed to create on these machine. I get:
```
JVMSHRC336E Port layer error code = -155
JVMSHRC337E Platform error message: EDC5124I Too many open files.
JVMSHRC840E Failed to start up the shared cache.
JVMJ9VM015W Initialization error for library j9shr29(11): JVMJ9VM009E J9VMDllMain failed
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

This means `mmap()` failed with `EMFILE`. Most SCC related tests directly fail on this error. 

I tried another build with the default cache size reduced to 10m, most of these SCC class loading tests passed. There are much fewer tests failed with the error above.

I guess there are machine settings that is related to this error. Probably things like `MAXFILEPROC`, `MAXMMAPAREA`, `MAXSHAREPAGES`. Could you check these settings on internal Jenkins z/OS machine, like fyrec51n.vmec.svl.ibm.com ? @jdekonin @vsebe

jdekonin commented:
```
bash-4.3$ oeconsol "D VS,HVSHARE"
IAR019I  10.28.37 DISPLAY VIRTSTOR 870
 SOURCE = DEFAULT
 TOTAL SHARED = 522240G
 SHARED RANGE = 2048G-524288G
 SHARED ALLOCATED = 1050446M
bash-4.3$ oeconsol "D OMVS,OPTIONS"
BPXO043I 10.27.08 DISPLAY OMVS 866
OMVS     000F ACTIVE             OMVS=(00)
CURRENT UNIX CONFIGURATION SETTINGS:
MAXPROCSYS      =        900    MAXPROCUSER     =        100
MAXFILEPROC     =      10000    MAXFILESIZE     = NOLIMIT
MAXCPUTIME      =      10800    MAXUIDS         =         50
MAXPTYS         =        800    MAXIOBUFUSER    =       2048
MAXMMAPAREA     =       4096    MAXASSIZE       = 2147483647
MAXTHREADS      =      10000    MAXTHREADTASKS  =      10000
MAXCORESIZE     =    4194304
IPCMSGQBYTES    = 2147483647    IPCMSGQMNUM     =      10000
IPCMSGNIDS      =        500    IPCSEMNIDS      =        500
IPCSEMNOPS      =         25    IPCSEMNSEMS     =       1000
IPCSHMMPAGES    =      25600    IPCSHMNIDS      =        500
IPCSHMNSEGS     =        500    IPCSHMSPAGES    =     262144
SUPERUSER       = BPXROOT       FORKCOPY        = COPY
STEPLIBLIST     =
USERIDALIASTABLE=
PRIORITYPG VALUES: NONE
PRIORITYGOAL VALUES: NONE
MAXQUEUEDSIGS   =       1000    SHRLIBRGNSIZE   =  157286400
SHRLIBMAXPAGES  =       4096    VERSION         = /       ,N
SYSCALL COUNTS  = NO            TTYGROUP        = TTY
SYSPLEX         = NO            BRLM SERVER     = N/A
LIMMSG          = SYSTEM        AUTOCVT         = OFF
RESOLVER PROC   = DEFAULT       LOSTMSG         = ON
AUTHPGMLIST     = NONE
SWA             = BELOW         NONEMPTYMOUNTPT = NOWARN
SERV_LINKLIB    =
SERV_LPALIB     =
MAXUSERMOUNTSYS =          0    MAXUSERMOUNTUSER=          0
MAXPIPEUSER     =       8730    PWT             = ENV
KERNELSTACKS    = ABOVE         UMASK           = NONE
```

hangshao0 commented:
I notice that I can create persistent SCC with a much larger size 60m on VM farm machine `stlab92.svl.ibm.com`:
```
java -Xshareclasses:name=C1,persistent,reset -Xscmx60m -version
JVMJ9VM082E Unable to switch to IFA processor - issue "extattr +a libj9ifa29.so"
JVMSHRC256I Persistent shared cache "C1" has been destroyed
java version "1.8.0_361"
Java(TM) SE Runtime Environment (build 8.0.8.0 - pmz6480sr8-20230208_02(SR8))
IBM J9 VM (build 2.9, JRE 1.8.0 z/OS s390x-64-Bit Compressed References 20230321_47772 (JIT enabled, AOT enabled)
OpenJ9   - cd369c6
OMR      - 0d8205c
IBM      - 103e724)
JCL - 20230208_01 based on Oracle jdk8u361-b09
```
Could you check the settings of `stlab92.svl.ibm.com` so that we can compare them against https://github.com/eclipse-openj9/openj9/pull/17073#issuecomment-1499399733 ? @jdekonin

jdekonin commented:
The 2 systems are not the same release.  fyrec51n is 2.5, whereas stlab92 is 2.1.  Note the older system has a MAXSHAREPAGES env var.
```
MAXPROCSYS      =       1000    MAXPROCUSER     =        500 
MAXFILEPROC     =      64000    MAXFILESIZE     = NOLIMIT    
MAXCPUTIME      =     100000    MAXUIDS         =        250 
MAXPTYS         =        512    MAXIOBUFUSER    =       2048 
MAXMMAPAREA     =      40960    MAXASSIZE       = 2147483647 
MAXTHREADS      =       1000    MAXTHREADTASKS  =       1000 
MAXCORESIZE     =    4194304    MAXSHAREPAGES   =     131072 
IPCMSGQBYTES    = 2147483647    IPCMSGQMNUM     =      10000 
IPCMSGNIDS      =        500    IPCSEMNIDS      =        500 
IPCSEMNOPS      =         25    IPCSEMNSEMS     =       1000 
IPCSHMMPAGES    =      25600    IPCSHMNIDS      =        500 
IPCSHMNSEGS     =        500    IPCSHMSPAGES    =     262144 
```

hangshao0 commented:
Looks like it is `MAXMMAPAREA` that is making a difference here: 

On `fyrec51n.vmec.svl.ibm.com`, it is 4096, which is equivalent to 16MB
On `stlab92.svl.ibm.com`,  it is 40960, which is equivalent to 160MB.

stlab92.svl.ibm.com can allow SCC with 160m, but fail with 161m.
```
java -Xshareclasses:name=C1,persistent,reset -Xscmx160m -version
JVMJ9VM082E Unable to switch to IFA processor - issue "extattr +a libj9ifa29.so"
JVMSHRC023E Cache does not exist
java version "1.8.0_361"
Java(TM) SE Runtime Environment (build 8.0.8.0 - pmz6480sr8-20230208_02(SR8))
IBM J9 VM (build 2.9, JRE 1.8.0 z/OS s390x-64-Bit Compressed References 20230321_47772 (JIT enabled, AOT enabled)
OpenJ9   - cd369c6
OMR      - 0d8205c
IBM      - 103e724)

java -Xshareclasses:name=C1,persistent,reset -Xscmx161m -version
JVMJ9VM082E Unable to switch to IFA processor - issue "extattr +a libj9ifa29.so"
JVMSHRC256I Persistent shared cache "C1" has been destroyed
JVMSHRC245E Error mapping shared class cache file
JVMSHRC336E Port layer error code = -155
JVMSHRC337E Platform error message: EDC5124I Too many open files. (errno2=0x07360344)
JVMSHRC840E Failed to start up the shared cache.
JVMJ9VM015W Initialization error for library j9shr29(11): JVMJ9VM009E J9VMDllMain failed
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

hangshao0 commented:
Seems that the `MAXMMAPAREA` value is too small on both VM farm and Jenkins z/OS test machines:

I just randomly tried on a VM farm machine `fyrec51a.vmec.svl.ibm.com`, 16m is allowed but 17m is too big:
```
java -Xshareclasses:name=C1,persistent,reset -Xscmx16m -version
JVMJ9VM082E Unable to switch to IFA processor - issue "extattr +a libj9ifa29.so"
JVMSHRC256I Persistent shared cache "C1" has been destroyed
java version "1.8.0_361"
Java(TM) SE Runtime Environment (build 8.0.8.0 - pmz6480sr8-20230208_02(SR8))
IBM J9 VM (build 2.9, JRE 1.8.0 z/OS s390x-64-Bit Compressed References 20230321_47772 (JIT enabled, AOT enabled)
OpenJ9   - cd369c6
OMR      - 0d8205c
IBM      - 103e724)
JCL - 20230208_01 based on Oracle jdk8u361-b09

java -Xshareclasses:name=C1,persistent,reset -Xscmx17m -version
JVMJ9VM082E Unable to switch to IFA processor - issue "extattr +a libj9ifa29.so"
JVMSHRC256I Persistent shared cache "C1" has been destroyed
JVMSHRC245E Error mapping shared class cache file
JVMSHRC336E Port layer error code = -155
JVMSHRC337E Platform error message: EDC5124I Too many open files. (errno2=0x07360344)
JVMSHRC840E Failed to start up the shared cache.
```

I guess we need to increase the `MAXMMAPAREA` value on z/OS machines. Maybe set it to 262,144 (==1GB) or 131,072 (==512MB) @jdekonin

pshipton commented:
Makes sense, however we should try to enable map64 by default, with appropriate documentation.

hangshao0 commented:
> however we should try to enable map64 by default, with appropriate documentation.

I agree it is better to use map64 whenever possible. @r30shah is talking with z/OS expert on the map64 failure that he saw likely due to machine configuration issues. I can turn map64 on by default once it is resolved.

r30shah commented:
On one of the development machine JIT team uses, we were able to get the `map64` option working. I needed to change the `SMFLIMxx` member in `SYS1.PARMLIB` and add following line,
```
REGION JOBNAME(*) MAXSHARE(2147483647)
```

And through SDSF, loaded the change with this command, 
```
/SET SMFLIM=00
```

With this change I was able to get `map64` with 2GB SCC working. I do not see this member in both of our perf box in order to get the performance results, Will try to see if we can request IT to update.

joransiu commented:
> Mentioned about turning map64 by default and fall back to below the bar map (with a warning message) if map64 failed.

@hangshao0 : In the default mode, if we are concerned about below-the-bar usage when system is not configured for map64, we can consider falling back to not using SCC with a warning highlight that's the case.

hangshao0 commented:
> In the default mode, if we are concerned about below-the-bar usage when system is not configured for map64, we can consider falling back to not using SCC with a warning highlight that's the case.

This is a behaviour when `-Xshareclass` sub-option `nonfatal` is used. If there's error starting up the SCC, it is considered non-fatal and the JVM continues to start up without using the SCC. A message on the SCC error will be printed out. I've seen many users using `nonfatal` if they want this behaviour. But `nonfatal` is not the default.

Another possibility is to check the cache size if `map64` failed. Fall back to below the bar map only if the SCC size is smaller than a threshold. We need to document this threshold size if we choose to do so.

r30shah commented:
Posting relative performance results compared to default SCC I got on z/OS (V2R4 8 CPs) running Daytrader 8 start-up runs with JDK11 on z/OS with configuration mentioned in https://github.com/eclipse-openj9/openj9/pull/17073#issuecomment-1503395875

| Configuration | Relative Performance |
|---------------|-----------------------|
| Baseline | 100 % |
| Persistent SCC | 103.1 % |
| Persistent SCC with Map64 | 101.9 % |

JVM Options used 
1. Baseline : `-Xcompressedrefs -Xms1024M -Xmx1024M -Xshareclasses -Xscmx128m`
2. Persistent SCC : `-Xcompressedrefs -Xms1024M -Xmx1024M -Xshareclasses:persistent -Xscmx128m`
3. Persistent SCC with Map64 : `-Xcompressedrefs -Xms1024M -Xmx1024M -Xshareclasses:persistent,map64`

hangshao0 commented:
>> however we should try to enable map64 by default, with appropriate documentation.

> I agree it is better to use map64 whenever possible. @r30shah is talking with z/OS expert on the map64 failure that he saw likely due to machine configuration issues. I can turn map64 on by default once it is resolved.

Change is updated to turn on map64 by default.

hangshao0 commented:
This can be reviewed again.

pshipton commented:
jenkins test sanity,extended win,zlinux jdk8,jdk11

keithc-ca commented on runtime/shared_common/OSCachemmap.cpp:
```diff
@@ -1663,13 +1666,19 @@ SH_OSCachemmap::syncUpdates(void* start, UDATA length, U_32 flags)
 	PORT_ACCESS_FROM_PORT(_portLibrary);
 
-	IDATA rc;
+	IDATA rc = 0;
```
This is C++ code, just move the declaration to line 1672 where `rc` gets its real value.

hangshao0 replied:
Fixed.

hangshao0 commented on runtime/nls/shrc/j9shr.nls:
```diff
@@ -6896,3 +6896,24 @@ J9NLS_SHRC_CM_PRINTSTATS_PROCESSOR_FEATURES.explanation=NOTAG
 J9NLS_SHRC_CM_PRINTSTATS_PROCESSOR_FEATURES.user_response=
 # END NON-TRANSLATABLE
+
+J9NLS_CC_MSYNC_UNSUPPORTED=The system does not support synchronizing memory with physical storage
```
Fixed.

hangshao0 commented on runtime/oti/j9.h:
```diff
@@ -375,6 +375,10 @@ static const struct { \
 #endif /* defined(OPENJ9_BUILD) */
 
+#if defined(J9VM_OPT_SHR_MSYNC_SUPPORT)
+#define	J9SHR_MSYNC_SUPPORT TRUE
```
> Please use a space instead of a tab after define.

Changed it to space.

> But why do we need/want this at all? We can use J9VM_OPT_SHR_MSYNC_SUPPORT directly.

We can replace the old macro `J9SHR_MSYNC_SUPPORT` with `J9VM_OPT_SHR_MSYNC_SUPPORT`. I don't want to make this PR bigger, so I didn't do that in this PR. It can be done in a separate PR (or separate commit), which could make it easier for reviewers to review.

keithc-ca replied:
This changes 7 out of the 8 files that used `J9SHR_MSYNC_SUPPORT` before; changing the 8th file instead of this makes the change smaller.

hangshao0 replied:
Removed `J9SHR_MSYNC_SUPPORT` and use `J9VM_OPT_SHR_MSYNC_SUPPORT` directly.

hangshao0 commented on runtime/shared_common/CompositeCache.cpp:
```diff
@@ -1396,32 +1381,61 @@ SH_CompositeCacheImpl::startup(J9VMThread* currentThread, J9SharedClassPreinitCo
+				if (isFirstStart) {
+					if ((0 == _theca->roundedPagesFlag)
+						|| _readOnlyOSCache
+					){
```
Fixed.

hangshao0 commented on runtime/shared_common/OSCachemmap.cpp:
```diff
@@ -1725,9 +1734,14 @@ SH_OSCachemmap::getPermissionsRegionGranularity(J9PortLibrary* portLibrary)
+#else
+	if (J9_ARE_ALL_BITS_SET(mmapCapabilities, J9PORT_MMAP_CAPABILITY_PROTECT))
+#endif /* defined(J9SHR_MSYNC_SUPPORT) */
+	{
```
Fixed.

keithc-ca replied:
The result is missing indentation before `if (...`.

hangshao0 replied:
Fixed.

hangshao0 replied:
>Suggest making a smaller part conditional:

```
        if (J9_ARE_ANY_BITS_SET(mmapCapabilities, J9PORT_MMAP_CAPABILITY_PROTECT
#if defined(J9VM_OPT_SHR_MSYNC_SUPPORT)
			| J9PORT_MMAP_CAPABILITY_MSYNC
#endif /* defined(J9VM_OPT_SHR_MSYNC_SUPPORT) */
	)) {
```

@keithc-ca This results in compilation errors:
```
"./OSCachemmap.cpp", line 1739.1: CCN5825 (W) The character "#" is not allowed.
"./OSCachemmap.cpp", line 1741.1: CCN5825 (W) The character "#" is not allowed.
"./OSCachemmap.cpp", line 1739.2: CCN5063 (S) The text "if" is unexpected.
```

keithc-ca replied:
Apparently the preprocessor isn't up to the challenge here (it seems to be having trouble with `#if` nested within the use of the macro `J9_ARE_ANY_BITS_SET`). New suggestion:
```c
#if defined(J9VM_OPT_SHR_MSYNC_SUPPORT)
#define MMAP_CAPABILITY_MASK (J9PORT_MMAP_CAPABILITY_PROTECT | J9PORT_MMAP_CAPABILITY_MSYNC)
#else /* defined(J9VM_OPT_SHR_MSYNC_SUPPORT) */
#define MMAP_CAPABILITY_MASK J9PORT_MMAP_CAPABILITY_PROTECT
#endif /* defined(J9VM_OPT_SHR_MSYNC_SUPPORT) */

	/* This call to capabilities is arguably unnecessary, but it is a good check to do */
	if (J9_ARE_ANY_BITS_SET(mmapCapabilities, MMAP_CAPABILITY_MASK)) {
		return j9mmap_get_region_granularity((void*)_headerStart);
	}
	return 0;

#undef MMAP_CAPABILITY_MASK
```

hangshao0 commented on runtime/tests/port/j9mmapTest.c:
```diff
@@ -179,6 +180,10 @@ j9mmap_test1(struct J9PortLibrary *portLibrary)
 
+	if (0 != extraMmapFlags) {
+		testName = "j9mmap_test1_extraFlags";
+	}
```
Fixed.

keithc-ca replied:
Same comment applies to `j9mmap_test2()`.

hangshao0 replied:
Fixed `j9mmap_test2()`.

hangshao0 commented on runtime/shared_common/shrinit.cpp:
```diff
@@ -277,6 +277,12 @@ J9SharedClassesHelpText J9SHAREDCLASSESHELPTEXT[] = {
 	{OPTION_DISABLE_CORRUPT_CACHE_DUMPS, 0, 0, J9NLS_SHRC_SHRINIT_HELPTEXT_DISABLE_CORRUPT_CACHE_DUMPS},
 	{OPTION_CHECK_STRINGTABLE_RESET, 0, 0, J9NLS_SHRC_SHRINIT_HELPTEXT_CHECK_STRINGTABLE_RESET},
 	{OPTION_ADDTESTJITHINT, 0, 0, J9NLS_SHRC_SHRINIT_HELPTEXT_ADD_JIT_HINTS},
+#if defined(J9ZOS39064) && (JAVA_SPEC_VERSION >= 11)
+	/* Flag __MAP_64 of mmap() is supported on z/OS version 2.4 and up. It is turned on by default on Java 11 and up.
+	 * option "map31" changes the mapping mode of Java 11 and up to 31-bit map.
+	 */
+	{OPTION_MAP31, 0, 0, J9NLS_SHRC_SHRINIT_HELPTEXT_MAP31},
+#endif /* defined(J9ZOS39064) && (JAVA_SPEC_VERSION >= 11) */
```
O/S version may not work, as __MAP_64 is supported on z/OS version 2.5 and on some of z/OS version 2.4. Version 2.4 needs to install certain update to have __MAP_64 support.

hangshao0 replied:
I am not aware of a way to detect if __MAP_64 is supported. We can use `zos_version_at_least()` to detect the z/OS version at runtime. The issue is we should detect z/OS 2.4 or 2.5. In our internal Jenkins testing machines, the 2.4 machines are updated to support ` __MAP_64`.

hangshao0 replied:
I notice we are compiling Java 11 and up on z/OS 2.4, so we should check for version 2.4. Change is updated to check version 2.4.

pshipton replied:
Java 11 z/OS only support 2.4 or later fyi.

hangshao0 replied:
On Java 11 and up, we will document the machine configuration needed for map64. We also need to document the z/OS update needed if users are on z/OS 2.4.

hangshao0 commented:
Any more review comments ?

pshipton commented on runtime/shared_common/OSCachemmap.cpp:
```diff
@@ -988,6 +991,13 @@ SH_OSCachemmap::internalAttach(bool isNewCache, UDATA generation)
 
+#if defined(J9ZOS39064)
+	if (J9_ARE_NO_BITS_SET(_runtimeFlags, J9SHR_RUNTIMEFLAG_MAP31)
+		&& zos_version_at_least(ZOS_V2R4_RELEASE, ZOS_V2R4_VERSION)
```
Hang, you are saying zos_version_at_least() is a compile time check?

Also I'm thinking we should fix OMR to allow __MAP_64 when running on V2.4 or later, even when compiling on V2.1. It's wrong to restrict Java 8 to below 2G when customers will be running on newer OSes that support __MAP_64. Versions earlier than V2.4 are already out of support, V2.3 EOS was Oct 2022.
https://github.com/eclipse/omr/blob/779c51b9568b18dbc6650c2ca284580e65aa4792/port/unix/omrmmap.c#L208

hangshao0 replied:
Is this change what you mean ?

pshipton replied:
It's fine for shared classes, but needs to behave correctly for OMR. As discussed on Slack, it depends on the behaviour of z/OS mmap when the flag is passed in. Since we don't have all the OSes 1.13 - 2.3 to test on, the OMR code should check for < 2.4 and return an error if OMRPORT_MMAP_FLAG_ZOS_64BIT is specified.

pshipton commented:
@keithc-ca this seems ready for another review. I want a change, but it's on the OMR side, and isn't a prereq for this.

hangshao0 commented:
The implementation of `zos_version_at_least()` is in OMR code, not in OpenJ9. I have removed its declaration from OpenJ9 header `util_api.h` in this PR and added it into OMR header` omrutil.h` in https://github.com/eclipse/omr/pull/6989

hangshao0 commented:
> I have removed its declaration from OpenJ9 header util_api.h in this PR and added it into OMR header omrutil.h in https://github.com/eclipse/omr/pull/6989

This creates dependency between the OMR and OpenJ9 PR.

keithc-ca commented on runtime/shared_common/shrinit.cpp:
```diff
@@ -376,6 +379,9 @@ J9SharedClassesOptions J9SHAREDCLASSESOPTIONS[] = {
 	{ OPTION_CREATE_LAYER, PARSE_TYPE_EXACT, RESULT_DO_CREATE_LAYER, 0 },
 #endif /* defined(J9VM_OPT_MULTI_LAYER_SHARED_CLASS_CACHE) */
 	{ OPTION_NO_PERSISTENT_DISK_SPACE_CHECK, PARSE_TYPE_EXACT, RESULT_DO_ADD_RUNTIMEFLAG, J9SHR_RUNTIMEFLAG_NO_PERSISTENT_DISK_SPACE_CHECK},
+#if (defined(J9ZOS39064)
```
Extra, unwanted `(` (I don't expect this to compile).

keithc-ca commented on runtime/tests/port/j9mmapTest.c:
```diff
@@ -264,15 +265,15 @@ j9mmap_test1(struct J9PortLibrary *portLibrary)
 	PORT_ACCESS_FROM_PORT(portLibrary);
-	const char* testName = "j9mmap_test2";
-
+	const char* testName = (0 == extraMmapFlags) ? "j9mmap_test2" : "j9mmap_test2_extraFlags";
```
Space before, not after `*`, please.

hangshao0 commented:
All existing review comments addressed.

keithc-ca commented:
Jenkins test sanity zlinux,win32 jdk8 depends https://github.com/eclipse/omr/pull/6989

keithc-ca reviewed (approved):
This will need to wait until https://github.com/eclipse/omr/pull/6989 is promoted to the openj9 branch of OMR.
